### PR TITLE
sys/shell: add XFA support

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds a macro that allows adding shell commands using a cross file array.

It basically allows this to be put in *any* file:

```
#include "shell.h"
static int _my_command(int argc, char **argv) {
 // ...
}

SHELL_COMMAND(my_command, "my command help text", _my_command);
```

### Testing procedure

This PR extends the tests/shell test scripts. Running this on some more platforms (AVR, RISC-V, ...) would be helpful!
Otherwise, please give feedback on the integration / usage.

### Issues/PRs references

First user of  [XFA](https://github.com/RIOT-OS/RIOT/pull/15002).

### CODE REVIEW

File: sys/include/shell.h [hunk 1]
```diff
@@ -202,6 +202,19 @@ static inline void shell_run(const shell_command_t *commands,
  *
  * This macro is a helper for defining a shell command and adding it to the
  * shell commands XFA (cross file array).
+ *
+ * @experimental This should be considered experimental API, subject to change
+ *               without notice!
```
Comment (miri64) [inline sys/include/shell.h:207]: The API seems simple enough, are there really changes expected to it? I could imagine that something might change in the XFA backend, but this shouldn't effect this macro as far as I can judge.

File: sys/include/shell.h [hunk 2]
```diff
@@ -21,9 +21,10 @@
 #define SHELL_H
 
 #include <stdint.h>
-#include "periph/pm.h"
 
 #include "kernel_defines.h"
+#include "periph/pm.h"
```
Comment (miri64) [inline sys/include/shell.h:26]: Seems like an unrelated style change. Maybe split this out into its own commit?

File: sys/shell/shell.c [hunk 3]
```diff
@@ -104,6 +121,10 @@ static shell_command_handler_t find_handler(
         handler = search_commands(_builtin_cmds, command);
     }
 
+    if (handler == NULL) {
```
Comment (miri64) [inline sys/shell/shell.c:124]: So if there is an `xfa` version of the command and a "classic" build-in version of the command, the classic version will be preferred? This should at the very least be documented.

File: sys/shell/shell.c [hunk 4]
```diff
@@ -104,6 +121,10 @@ static shell_command_handler_t find_handler(
...
+        handler = search_commands_xfa(command);
+    }
```
Comment (miri64) [inline sys/shell/shell.c:126]: Should `shell_command_xfa` maybe be a pseudo-module, so we are able to switch of this experimental code in production?

File: sys/include/shell.h (hunk 1 above)
Comment (kaspar030) [reply sys/include/shell.h:207]: True, but as XFA is experimental, I'd rather not have external users depend on this for now. Time will tell if issues pop up with XFA in general...

Also, there might be a way to deduce the function name from the command name (e.g., by requiring the function to be called `_shell_command_<name>`). I'd like to be able to do that without going through API change procedures.

**Review by miri64 — CHANGES_REQUESTED** [review]

File: sys/include/shell.h (hunk 1 above)
Comment (miri64) [reply sys/include/shell.h:207]: > True, but as XFA is experimental, I'd rather not have external users depend on this for now. Time will tell if issues pop up with XFA in general...

This can be experimental without stating that the API is subject to change ;-).



> Also, there might be a way to deduce the function name from the command name (e.g., by requiring the function to be called `_shell_command_<name>`). I'd like to be able to do that without going through API change procedures.

I would rather make this a convenience function in addition to `SHELL_COMMAND()`. There is no technical reason for the user to not keep the freedom to call the handler whatever they want it to be called.

File: sys/include/shell.h (hunk 2 above)
Comment (kaspar030) [reply sys/include/shell.h:26]: hm I'm lazy and would rather save the reset/commit/commit cycle.

File: sys/shell/shell.c (hunk 3 above)
Comment (kaspar030) [reply sys/shell/shell.c:124]: hmya. same with built-in and user specified. Is that documented? If not, that's a doc buf for a followup.

File: sys/shell/shell.c (hunk 4 above)
Comment (kaspar030) [reply sys/shell/shell.c:126]: that sounds useful, will do

File: sys/include/shell.h (hunk 1 above)
Comment (maribu) [reply sys/include/shell.h:207]: > There is no technical reason for the user to not keep the freedom to call the handler whatever they want it to be called.

It would simplify searching the source code significantly, if a simple regex could detect any shell command implementation.

File: sys/include/shell.h (hunk 2 above)
Comment (maribu) [reply sys/include/shell.h:26]: Make a fixup commit this undoes this, than create a revert commit of the fixup. Then keep the reverting commit and squash the fixup in.

It would be nice to have a vim plugin to automate this.

File: sys/shell/shell.c (hunk 4 above)
Comment (kaspar030) [reply sys/shell/shell.c:131]: done

File: sys/include/shell.h (hunk 2 above)
Comment (kaspar030) [reply sys/include/shell.h]: ok, style fix reverted

File: sys/shell/shell.c (hunk 3 above)
Comment (miri64) [reply sys/shell/shell.c:129]: No it's not, but this is literally just a note you can stick to the doc `SHELL_COMMAND`. The doc for build-ins, yes, that can be done as a follow-up.

File: sys/include/shell.h (hunk 1 above)
Comment (benpicco) [reply sys/include/shell.h:209]: >  there might be a way to deduce the function name from the command name (e.g., by requiring the function to be called `_shell_command_<name>`

I think this is one of those instances where the convenience of the auto-magic does not justify the added complexity of making it non-obvious what's going on.

File: sys/shell/shell.c (hunk 3 above)
Comment (kaspar030) [reply sys/shell/shell.c:129]: done

File: sys/include/shell.h [hunk 5]
```diff
@@ -196,6 +197,34 @@ static inline void shell_run(const shell_command_t *commands,
...
+ *
+ * ```.c
+ * #include "shell.h"
+ * static int _my_command(int argc, char **argv) {
+ *   // ...
+ * }
+ * SHELL_COMMAND(my_command, "my command help text", _my_command);
+ * ```
+ */
+#if defined(DOXYGEN) || defined(MODULE_SHELL_COMMAND_XFA)
```
Comment (miri64) [inline sys/include/shell.h:221]:
```suggestion
#if defined(DOXYGEN) || IS_USED(MODULE_SHELL_COMMAND_XFA)
```

We should finally add a coccinelle check for that...

File: sys/shell/shell.c [hunk 6]
```diff
@@ -36,9 +36,15 @@
 #include <errno.h>
 
 #include "kernel_defines.h"
+#include "xfa.h"
 #include "shell.h"
 #include "shell_commands.h"
 
+#if defined(MODULE_SHELL_COMMAND_XFA)
```
Comment (miri64) [inline sys/shell/shell.c:43]:
```suggestion
#if IS_USED(MODULE_SHELL_COMMAND_XFA)
```

File: sys/include/shell.h (hunk 1 above)
Comment (kaspar030) [reply sys/include/shell.h:209]: > This can be experimental without stating that the API is subject to change ;-).

It is our standard disclaimer for new APIs...

File: sys/shell/shell.c [hunk 7]
```diff
@@ -104,6 +125,12 @@ static shell_command_handler_t find_handler(
         handler = search_commands(_builtin_cmds, command);
     }
 
+#if defined(MODULE_SHELL_COMMAND_XFA)
+    if (handler == NULL) {
+        handler = search_commands_xfa(command);
+    }
+#endif
```
Comment (miri64) [inline sys/shell/shell.c:132]:
```suggestion
    if (IS_USED(MODULE_SHELL_COMMAND_XFA) && (handler == NULL)) {
        handler = search_commands_xfa(command);
    }
```

File: sys/shell/shell.c [hunk 8]
```diff
@@ -114,6 +141,17 @@ static void print_commands(const shell_command_t *entry)
...
+#if defined(MODULE_SHELL_COMMAND_XFA)
+static void print_commands_xfa(void)
+{
+    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
+    for (unsigned i = 0; i < n; i++) {
+        const volatile shell_command_t *entry = shell_commands_xfa[i];
+        printf("%-20s %s\n", entry->name, entry->desc);
+    }
+}
+#endif
```
Comment (miri64) [inline sys/shell/shell.c:153]:
```suggestion
static void print_commands_xfa(void)
{
#if IS_USED(MODULE_SHELL_COMMAND_XFA)
    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
    for (unsigned i = 0; i < n; i++) {
        const volatile shell_command_t *entry = shell_commands_xfa[i];
        printf("%-20s %s\n", entry->name, entry->desc);
    }
#endif
}
```

File: sys/shell/shell.c [hunk 9]
```diff
@@ -125,6 +163,10 @@ static void print_help(const shell_command_t *command_list)
     if (_builtin_cmds != NULL) {
         print_commands(_builtin_cmds);
     }
+
+#if defined(MODULE_SHELL_COMMAND_XFA)
+    print_commands_xfa();
+#endif
```
Comment (miri64) [inline sys/shell/shell.c:169]:
```suggestion
    if (IS_USED(MODULE_SHELL_COMMAND_XFA)) {
        print_commands_xfa();
    }
```

**Review by miri64 — DISMISSED** [review]

File: sys/include/shell.h (hunk 1 above)
Comment (miri64) [reply sys/include/shell.h:209]: Ok, I did not know that we have a standard disclaimer.

File: sys/shell/shell.c (hunk 3 above)
Comment (miri64) [reply sys/shell/shell.c:129]: Can you provide a link? I can't find it.

**Comment by kaspar030** [comment]
I merged the IS_USED suggestions (they're in a fixup now).

**Comment by kaspar030** [comment]
Somehow the IS_USED make this a bit messy, as the top-level definition is either always there or not available for within the IS_USED blocks. @miri64 do you mind if I revert that?

**Comment by miri64** [comment]
> Somehow the IS_USED make this a bit messy, as the top-level definition is either always there or not available for within the IS_USED blocks. @miri64 do you mind if I revert that?

I am not sure what you mean by "top-level definition" or how `IS_USED()` makes things more messy than using `defined()` or `#ifdef`.

File: sys/shell/shell.c [hunk 10]
```diff
@@ -92,6 +98,21 @@ static shell_command_handler_t search_commands(const shell_command_t *entry,
...
+
+    for (unsigned i = 0; i < n; i++) {
+        const volatile shell_command_t *entry = shell_commands_xfa[i];
+        if (strcmp(entry->name, command) == 0) {
+            return entry->handler;
+        }
+    }
+    return NULL;
+}
+#endif
```
Comment (miri64) [inline sys/shell/shell.c:114]: This needs to be

```suggestion
static shell_command_handler_t search_commands_xfa(char *command)
{
#if IS_USED(MODULE_SHELL_COMMAND_XFA)
    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);

    for (unsigned i = 0; i < n; i++) {
        const volatile shell_command_t *entry = shell_commands_xfa[i];
        if (strcmp(entry->name, command) == 0) {
            return entry->handler;
        }
    }
    return NULL;
#else
    (void)command;
#endif
}
```

File: sys/shell/shell.c [hunk 11]
```diff
@@ -104,6 +125,10 @@ static shell_command_handler_t find_handler(
         handler = search_commands(_builtin_cmds, command);
     }
 
+    if (IS_USED(MODULE_SHELL_COMMAND_XFA) && (handler == NULL)) {
+        handler = search_commands_xfa(command);
+    }
```
Comment (miri64) [inline sys/shell/shell.c:130]: ... for this to work.

**Review by miri64 — COMMENTED** [review]
Do you mean this?

File: sys/shell/shell.c (hunk 10 above)
Comment (miri64) [inline sys/shell/shell.c:114]:
```suggestion
#if IS_USED(MODULE_SHELL_COMMAND_XFA)
static shell_command_handler_t search_commands_xfa(char *command)
{
    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);

    for (unsigned i = 0; i < n; i++) {
        const volatile shell_command_t *entry = shell_commands_xfa[i];
        if (strcmp(entry->name, command) == 0) {
            return entry->handler;
        }
    }
    return NULL;
}
#endif
```

File: sys/shell/shell.c (hunk 11 above)
Comment (miri64) [inline sys/shell/shell.c:130]:
```suggestion
#if IS_USED(MODULE_SHELL_COMMAND_XFA)
    if (handler == NULL) {
        handler = search_commands_xfa(command);
    }
#endif
```

**Review by miri64 — COMMENTED** [review]
Alternatively, this would also work, if that is less messy in your opinion but still be within the coding conventions.

File: sys/shell/shell.c (hunk 10 above)
Comment (kaspar030) [reply sys/shell/shell.c]: missing return in else case

Comment (miri64) [reply sys/shell/shell.c]: Ah, yeah, the `return NULL` needs to be outside of the `#if` block, but I see you already did that :-).

**Comment by kaspar030** [comment]
> I am not sure what you mean by "top-level definition" or how `IS_USED()` makes things more messy than using `defined()` or `#ifdef`.

I mean this:

```
 43 #if IS_USED(MODULE_SHELL_COMMAND_XFA)                                                                                          
 44 /* define shell command cross file array */                                                                                     45 XFA_INIT_CONST(shell_command_t*, shell_commands_xfa);                                                                          
 46 #endif                                                                                                                         
```

> I am not sure what you mean by "top-level definition" or how `IS_USED()` makes things more messy than using `defined()` or `#ifdef`.

I had used C if instead of preprocessor if whithin a function, which failed because it was missing a file-level definition (which is also behind preprocessor #if).

**Comment by miri64** [comment]
> I am a bit wary about an optional feature extension (`shell_commands_xfa`) being added to the main test for the base feature (`shell`), but it is not a blocking thing for me. What do @HendrikVE or @fjmolinas (as the people who last worked on that test) think about this?

To make things perfectly clear, this is just about moving the tests introduced here to their own application.

**Comment by kaspar030** [comment]
> To make things perfectly clear, this is just about moving the tests introduced here to their own application.

"just"

How about we make this un-optional again. This is supposed to be a huge improvement in how we organize code. We don't expect any problems. I tagged this "experimental" because XFA is kinda new and _the API might change_. I don't want to dance around this by IS_USED and a seperate test application.

@miri64 do you see something wrong with the general approach?

If not, again, I'd rather have this non-optional and get maximum exposure (fail fast), so we quickly see if the approach is viable.

**Comment by miri64** [comment]
> How about we make this un-optional again. This is supposed to be a huge improvement in how we organize code. We don't expect any problems. I tagged this "experimental" because XFA is kinda new and _the API might change_. I don't want to dance around this by IS_USED and a seperate test application.

Again, I don't mind it really to be in there. It just is my inner alarm bells ringing adding an (optional) experimental feature to what seems to me like a very central test (after all the shell is very important to our testing).
 
> @miri64 do you see something wrong with the general approach?

Nope.

> If not, again, I'd rather have this non-optional and get maximum exposure (fail fast), so we quickly see if the approach is viable.

Again, IMHO this can be merged as is, I just want some more input by people who worked with the shell more in-depth. If there is no problem from there side or no input, this can go in ASAP.

**Comment by kaspar030** [comment]
again, I'd rather have this non-optional and get maximum exposure (fail fast), so we quickly see if the approach is viable.

Let's not hide this but expose it to the max, so we can have confidence at the end of the release cycle that the approach is viable.

**Comment by chrysn** [comment]
While just familiarizing myself with XFA I support this being nonoptional. Fewer pseudomodules, less ifdef, and if XFA does as I remember it being advertised, as long as nothing hooks in it's a no-op (as in absent, not as in the ASM instruction) when built.

**Comment by kaspar030** [comment]
> and if XFA does as I remember it being advertised, as long as nothing hooks in it's a no-op (as in absent, not as in the ASM instruction) when built.

Hm, did we advertize that? I think the loops iterating XFAs will not optimize out, as the array lengths are determined at runtime.

**Comment by miri64** [comment]
>  Fewer pseudomodules, less ifdef, and if XFA does as I remember it being advertised, as long as nothing hooks in it's a no-op (as in absent, not as in the ASM instruction) when built.

Mh... but doesn't this make it problematic when a user expects it to be there, but it isn't? With a pseudomodule we have at least a chance to warn the user about missing functionality.

**Comment by miri64** [comment]
> Let's not hide this but expose it to the max, so we can have confidence at the end of the release cycle that the approach is viable.

And what if it is not? Do we hastily revert everything regarding XFAs then?

**Comment by chrysn** [comment]
> With a pseudomodule we have at least a chance to warn the user about missing functionality.

By that reasoning we'd have a pseudomodule for everything that's introduced inbetween releases and may go away again. I think that'd be excessive versioning, unless we have a way of listing features as "It's OK to require it but really you don't as it's been built-in for releases".

**Comment by chrysn** [comment]
> I think the loops iterating XFAs will not optimize out, as the array lengths are determined at runtime.

Not sure, but that's the general impression I got from XFA. But indeed it is not the case (tested, and even with LTO, XFA_LEN is still calculated at runtime from comparing two loaded values that happen to be identical). Sorry for the uninformed noise.

File: sys/shell/shell.c [hunk 12]
```diff
@@ -92,6 +98,23 @@ static shell_command_handler_t search_commands(const shell_command_t *entry,
     return NULL;
 }
 
+static shell_command_handler_t search_commands_xfa(char *command)
+{
+#if IS_USED(MODULE_SHELL_COMMAND_XFA)
+    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
+
+    for (unsigned i = 0; i < n; i++) {
+        const volatile shell_command_t *entry = shell_commands_xfa[i];
```
Comment (maribu) [inline sys/shell/shell.c:107]: Why `volatile` here?

File: sys/shell/shell.c [hunk 13]
```diff
@@ -114,6 +141,17 @@ static void print_commands(const shell_command_t *entry)
     }
 }
 
+static void print_commands_xfa(void)
+{
+#if IS_USED(MODULE_SHELL_COMMAND_XFA)
+    unsigned n = XFA_LEN(shell_command_t*, shell_commands_xfa);
+    for (unsigned i = 0; i < n; i++) {
+        const volatile shell_command_t *entry = shell_commands_xfa[i];
```
Comment (maribu) [inline sys/shell/shell.c:149]: Same

File: tests/shell/main.c [hunk 14]
```diff
@@ -103,6 +103,30 @@ static const shell_command_t shell_commands[] = {
...
+    (void) argv;
+    printf("[XFA TEST 2 OK]\n");
+
+    return 0;
+}
+
+/* Add above commands to the shell commands XFA using helper macro.
+ * Intentionally reversed order to test linker script based alphanumeric
+ * ordering. */
+SHELL_COMMAND(xfa_test2, "xfa test command 2",_xfa_test2);
```
Comment (maribu) [inline tests/shell/main.c:127]:
```suggestion
SHELL_COMMAND(xfa_test2, "xfa test command 2", _xfa_test2);
```

File: tests/shell/main.c [hunk 15]
```diff
@@ -103,6 +103,30 @@ static const shell_command_t shell_commands[] = {
...
+SHELL_COMMAND(xfa_test1, "xfa test command 1",_xfa_test1);
```
Comment (maribu) [inline tests/shell/main.c:128]:
```suggestion
SHELL_COMMAND(xfa_test1, "xfa test command 1", _xfa_test1);
```

File: tests/shell/main.c [hunk 16]
```diff
@@ -103,6 +103,30 @@ static const shell_command_t shell_commands[] = {
     { NULL, NULL, NULL }
 };
 
+static int _xfa_test1(int argc, char **argv)
+{
+    (void) argc;
+    (void) argv;
```
Comment (maribu) [inline tests/shell/main.c:109]:
```suggestion
    (void)argc;
    (void)argv;
```

File: tests/shell/main.c [hunk 17]
```diff
@@ -103,6 +103,30 @@ static const shell_command_t shell_commands[] = {
...
+    printf("[XFA TEST 1 OK]\n");
+
+    return 0;
+}
+
+static int _xfa_test2(int argc, char **argv)
+{
+    (void) argc;
+    (void) argv;
```
Comment (maribu) [inline tests/shell/main.c:118]:
```suggestion
    (void)argc;
    (void)argv;
```

**Review by maribu — COMMENTED** [review]
Looks good to me. Some comment inline.

File: tests/shell/main.c (hunk 14 above)
Comment (kaspar030) [reply tests/shell/main.c:127]: I've uncrustified the file, fixing all the style remarks

**Comment by kaspar030** [comment]
> Mh... but doesn't this make it problematic when a user expects it to be there, but it isn't? With a pseudomodule we have at least a chance to warn the user about missing functionality.

XFA is always there ;)

**Comment by kaspar030** [comment]
> And what if it is not? Do we hastily revert everything regarding XFAs then?

If it comes to that, yes. For that we need to expose it and use it in some significant way. I don't _expect_ problems.

I marked XFA (and this PR's added macro) experimental so external users don't jump on it while it might still stabilize.

**Comment by miri64** [comment]
Mhh... okay, if you think this works out better in the end (I think that too, if XFAs work as advertised) and if this makes things easier, we can revert the pseudo-module changes. Especially, since XFAs themselves are not optional either.

**Comment by fjmolinas** [comment]
> This PR extends the tests/shell test scripts. Running this on some more platforms (AVR, RISC-V, ...) would be helpful!
> Otherwise, please give feedback on the integration / usage.

I'll test with many platforms, but no MIPS or ESP.

**Comment by kaspar030** [comment]
> Mhh... okay, if you think this works out better in the end (I think that too, if XFAs work as advertised) and if this makes things easier, we can revert the pseudo-module changes. Especially, since XFAs themselves are not optional either.

thanks! I promise to re-optionalize should it prove to be a valuable optimization!

**Comment by kaspar030** [comment]
> ESP

CI can do that.

**Comment by fjmolinas** [comment]
@kaspar030 soemthing si off application does not compile

**Comment by fjmolinas** [comment]
```
In file included from /home/francisco/workspace/RIOT/sys/include/shell.h:27:0,
                 from /home/francisco/workspace/RIOT/sys/include/shell_commands.h:24,
                 from /home/francisco/workspace/RIOT/tests/shell/main.c:24:
/home/francisco/workspace/RIOT/sys/include/shell.h:231:49: error: expected declaration specifiers or '...' before '&' token
     XFA_ADD_PTR(shell_commands_xfa, name, name, &_xfa_ ## name ## _cmd)
                                                 ^
/home/francisco/workspace/RIOT/core/include/xfa.h:168:18: note: in definition of macro 'XFA_ADD_PTR'
     const typeof(entry) xfa_name ## _ ## prio ## _ ## name = entry
                  ^~~~~
/home/francisco/workspace/RIOT/tests/shell/main.c:127:1: note: in expansion of macro 'SHELL_COMMAND'
 SHELL_COMMAND(xfa_test2, "xfa test command 2", _xfa_test2);
 ^~~~~~~~~~~~~
/home/francisco/workspace/RIOT/sys/include/shell.h:231:49: error: expected declaration specifiers or '...' before '&' token
     XFA_ADD_PTR(shell_commands_xfa, name, name, &_xfa_ ## name ## _cmd)
                                                 ^
/home/francisco/workspace/RIOT/core/include/xfa.h:168:18: note: in definition of macro 'XFA_ADD_PTR'
     const typeof(entry) xfa_name ## _ ## prio ## _ ## name = entry
                  ^~~~~
/home/francisco/workspace/RIOT/tests/shell/main.c:128:1: note: in expansion of macro 'SHELL_COMMAND'
 SHELL_COMMAND(xfa_test1, "xfa test command 1", _xfa_test1);
 ^~~~~~~~~~~~~
In file included from /home/francisco/workspace/RIOT/sys/include/shell_commands.h:24:0,
                 from /home/francisco/workspace/RIOT/tests/shell/main.c:24:
/home/francisco/workspace/RIOT/sys/include/shell.h:230:28: error: '_xfa_xfa_test1_cmd' defined but not used [-Werror=unused-variable]
     static shell_command_t _xfa_ ## name ## _cmd = { #name, help, &func }; \
                            ^
/home/francisco/workspace/RIOT/tests/shell/main.c:128:1: note: in expansion of macro 'SHELL_COMMAND'
 SHELL_COMMAND(xfa_test1, "xfa test command 1", _xfa_test1);
 ^~~~~~~~~~~~~
/home/francisco/workspace/RIOT/sys/include/shell.h:230:28: error: '_xfa_xfa_test2_cmd' defined but not used [-Werror=unused-variable]
     static shell_command_t _xfa_ ## name ## _cmd = { #name, help, &func }; \
                            ^
/home/francisco/workspace/RIOT/tests/shell/main.c:127:1: note: in expansion of macro 'SHELL_COMMAND'
 SHELL_COMMAND(xfa_test2, "xfa test command 2", _xfa_test2);

```

**Comment by kaspar030** [comment]
> @kaspar030 soemthing si off application does not compile

yeah, here too. looking into it.

File: sys/shell/shell.c (hunk 12 above)
Comment (maribu) [reply sys/shell/shell.c:104]: Hmm, sounds a bit like a foot gun, as it is easy to forget.

Also, this marks the memory `entry` points to as `volatile`. So very access to a member of `entry` will result in a read from memory, rather than accessing a register that is still holding the value. This is not an issue here with strings, but it is not too hard to imagine situations in which this would result in poorly optimized code.

I think that it might be that the compiler only falsely assumes that the length is zero. There are some bugs in GCC leading to unoptimised code surrounding `volatile` accesses. If this is the case, newer versions of GCC without this bug would still optimize out the loop. I will try to find some time to dig into this and more specifically tell the compiler where not to optimize.

File: core/include/xfa.h [hunk 18]
```diff
@@ -158,14 +158,20 @@ _Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
  *
  * Pointers will end up sorted by prio.
  *
+ * @note This implementation uses the __typeof__() C extension.
+ *       It is available both in GCC and LLCM, and both don't complain with
+ *       plain "-std=c11". If deemed necessary, use of __typeof__ can be removed
+ *       and the type can be added as parameter, at the cost of much less
+ *       convenience.
```
Comment (miri64) [inline core/include/xfa.h:165]: Usage of `__typeof__` vs `typeof` could also be made a compile-time switch. However, since `embUnit` is also using `__typeof__` and we never had a problem with that, I don't see any problems coming our way.

File: core/include/xfa.h [hunk 19]
```diff
@@ -158,14 +158,20 @@ _Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
...
+ *
  * @param[in]   xfa_name    name of the xfa
  * @param[in]   prio        priority within the xfa
  * @param[in]   name        symbol name
  * @param[in]   entry       pointer variable to add to xfa
  */
```
Comment (miri64) [inline core/include/xfa.h:171]: See https://github.com/RIOT-OS/RIOT/pull/15795 to get proper fix proposals here ;-)

File: core/include/xfa.h [hunk 20]
```diff
@@ -158,14 +158,20 @@ _Pragma("GCC diagnostic ignored \"-Warray-bounds\"")
  *
  * Pointers will end up sorted by prio.
  *
+ * @note This implementation uses the __typeof__() C extension.
+ *       It is available both in GCC and LLCM, and both don't complain with
```
[the remark this answers is not in the document]
Comment (kaspar030) [reply core/include/xfa.h:162]: yup.

File: sys/shell/shell.c (hunk 12 above)
Comment (kaspar030) [reply sys/shell/shell.c:104]: > Hmm, sounds a bit like a foot gun, as it is easy to forget.

what would be the footgun? gcc does assume the array has length 0 (it is defined as such ...).
The volatile is enforced by the compiler's type check (and by the XFA macros), it should not be possible to accidentally not use volatilr.

The missed optimization opportunities are a thing. I'd hope they's not be so bad...

File: sys/shell/shell.c (hunk 3 above)
Comment (kaspar030) [reply sys/shell/shell.c:124]: I stuck a note to `SHELL_COMMAND()`:
https://github.com/kaspar030/RIOT/blob/a36f9ce7744edba9fa473a9a9d7a5aab2953c827/sys/include/shell.h#L206-L209

**Comment by fjmolinas** [comment]
@kaspar030 is this ready for a test?

File: core/include/xfa.h (hunk 19 above)
Comment (kaspar030) [inline core/include/xfa.h:171]: mmmh my uncrustify disagrees. this is a warning still, right?

**Comment by kaspar030** [comment]
Does it make sense to split out the core/include/xfa.h commit?

**Comment by kaspar030** [comment]
> @kaspar030 is this ready for a test?

yes! (just realized I didn't send a heads up comment).

**Comment by kaspar030** [comment]
> Does it make sense to split out the core/include/xfa.h commit?

I did. https://github.com/RIOT-OS/RIOT/pull/16076
This PR contains that change still, will rebase once that one is in.

File: sys/shell/shell.c (hunk 12 above)
Comment (maribu) [reply sys/shell/shell.c:104]: > The volatile is enforced by the compiler's type check (and by the XFA macros), it should not be possible to accidentally not use volatilr.

OK, no footgun then :-)

**Comment by fjmolinas** [comment]
All good test wise https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/242/artifact/.

**Comment by fjmolinas** [comment]
please rebase @kaspar030!

**Comment by kaspar030** [comment]
Unrelated tests failing:

```
--- run_test job results (1 failed, 211 passed, 212 total):
    failed:
    tests/mtd_raw/esp32-wroom-32:gnu
```

application "tests_periph_spi_dma" for "samd10-xmini" failed as it got too large. investigating.

**Comment by kaspar030** [comment]
> application "tests_periph_spi_dma" for "samd10-xmini" failed as it got too large. investigating.

So this does enlarge shell users by 120b, which is IMO too much if unused.

I'll try converting the internal commands to use XFA and see if that mitigates it. If yes, nice, if not, I'll make this optional again.

**Comment by kaspar030** [comment]
> I'll try converting the internal commands to use XFA and see if that mitigates it. If yes, nice, if not, I'll make this optional again.

It does somewhat (size increase is 40b instead of 120b). I'll open a followup for that change.

**Comment by kaspar030** [comment]
I realized I didn't make the shell commands defined by `SHELL_COMMANDS()` "const", so they ended up in RAM. Latest commit fixes that.

**Comment by kaspar030** [comment]
- squashed

@fjmolinas would you give this another test run? I added a "const" in the macro which should get some testing...

**Comment by fjmolinas** [comment]
>     * squashed
> 
> 
> @fjmolinas would you give this another test run? I added a "const" in the macro which should get some testing...

All good, I had a failure on `nucleo-l433rc` but that failures is in master as well.

File: tests/shell/main.c [hunk 21]
```diff
@@ -121,8 +145,8 @@ int main(void)
 
     /* or use only system shell commands */
     /*
-    shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE);
-    */
+       shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE);
+     */
```
Comment (miri64) [inline tests/shell/main.c:149]: Nit-pick: [`vim-commentary`](https://github.com/tpope/vim-commentary) doesn't like either format. The previous format at least allowed for only removing the `/*` and `*/` to have clean code. An `uncrustrify`-compliant way that `vim-commentary` would also be happy with could be

```C
    /* shell_run(NULL, line_buf, SHELL_DEFAULT_BUFSIZE); */
```

**Review by miri64 — COMMENTED** [review]
Only a minor comment left. Please squash.

File: tests/shell/main.c (hunk 21 above)
Comment (kaspar030) [reply tests/shell/main.c]: fixed

**Review by miri64 — APPROVED** [review]
ACK, this PR was [previously tested](https://github.com/RIOT-OS/RIOT/pull/16061#issuecomment-784218514) by @fjmolinas and [retested](https://github.com/RIOT-OS/RIOT/pull/16061#issuecomment-789666629) after the latest rework. If Murdock is happy, this can go in.

**Comment by chrysn** [comment]
Just for better understanding this: Why are the pointers to the command structs kept in XFA, and nit the command structs right away? Seems to me like a useless indirection, but maybe I missed the reason in the discussion.

**Comment by kaspar030** [comment]
> Just for better understanding this: Why are the pointers to the command structs kept in XFA, and nit the command structs right away? Seems to me like a useless indirection, but maybe I missed the reason in the discussion.

hm, I didn't think much about this. I think both are fine. keeping just pointers in XFA makes the size of the shell command thingie show up together with the .data of the module defining it. And, the xfa itself has to be volatile, whereas the data pointed to doesn't.

Would changing this make anything easier?

**Comment by chrysn** [comment]
> hm, I didn't think much about this. I think both are fine.

Yeah, it's a small detail; my main concern was for the increased memory
use and indirection in iteration, but that's, like, 1 pointer per
command and 1 deref instruction.

> keeping just pointers in XFA makes the size of the shell command
> thingie show up together with the .data of the module defining it.

I think it still would if the full shell command struct were in XFA. (nm
& co count symbol sizes, and each entry in the XFA is an own symbol with
inside some object file with its proper source attribution).

> And, the xfa itself has to be volatile, whereas the data pointed to
> doesn't.

OK, the volatility makes sense because it's part of the type. (It's
different in Rust so I didn't notice).

> Would changing this make anything easier?

No, it would just shave, like, 3 lines off the wrappers.
